src/transformers/models/jukebox/modeling_jukebox.py

```diff
@@ -0,0 +1,3133 @@
+        self.width = np.prod(normalized_shape)
+        self.max_numel = 65535 * self.width
+
+    def forward(self, input):
+        if input.numel() > self.max_numel:
+            return F.layer_norm(input, self.normalized_shape, self.weight, self.bias, self.eps).type_as(input)
+        else:
+            return super(JukeboxLayerNorm, self).forward(input).type_as(input)
+
+
+def get_mask(mask, query_length, key_value_length, blocks, spread, device, sample, sample_t):
+    # returns a mask of shape 1 x 1 x query_length x key_value_length or None if masking is not needed.
+    if mask is None or query_length == 1:
+        return None
+    offset = sample_t - query_length if sample else max(key_value_length - query_length, 0)
+    if mask == "autoregressive":
+        # Masked dense
+        mask = torch.ones(query_length, key_value_length, device=device).tril(offset)
+    elif mask == "summary":
+        # Masked summary
+        mask = (
+            torch.nn.functional.pad(
+                torch.ones(query_length, query_length, device=device)
+                .tril()
+                .view(query_length, blocks, query_length // blocks)[:, :-1, -key_value_length // blocks :],
+                (0, 0, 1, 0),
+                value=1,
+            )
+            .contiguous()
+            .view(query_length, key_value_length)
+        )
+    elif mask == "prime":
+        mask = torch.ones(query_length, key_value_length, device=device).tril(offset)
+    return mask.view(1, 1, query_length, key_value_length)
+
+
+class JukeboxAttention(nn.Module):
+    def __init__(
+        self,
+        width,
+        n_ctx,
+        hidden_dim,
+        num_heads,
+        attn_dropout=0.0,
+        resid_dropout=0.0,
+        scale=True,
+        mask=False,
+        zero_out=False,
+        init_scale=1.0,
+        attn_func=0,
+        blocks=None,
+        spread=None,
+        encoder_dims=None,
+        lyric_enc_len=None,
+    ):
+        super().__init__()
+        self.width = width  # should have a better name
+        self.n_ctx = n_ctx  # NOTE: n_ctx could be different within operations. This is complete n_ctx
+        self.hidden_dim = hidden_dim
+        self.num_heads = num_heads
+        self.scale = scale
+        self.mask = mask
+        if attn_func == 6:
+            self.c_attn = JukeboxConv1D(width, hidden_dim)
+            self.c_enc_kv = JukeboxConv1D(width, hidden_dim * 2)
+        else:
+            self.c_attn = JukeboxConv1D(width, hidden_dim * 3)
+        self.c_proj = JukeboxConv1D(hidden_dim, width, zero_out)
+        self.attn_dropout = nn.Dropout(attn_dropout) if attn_dropout > 0.0 else lambda x: x
+        self.resid_dropout = nn.Dropout(resid_dropout) if resid_dropout > 0.0 else lambda x: x
+
+        # Sequence of length seq_len is factored as [blocks, seq_len // blocks]
+        self.attn_func = attn_func
+        self.qkv, self.attn, self.attn_mask = {
```

Comment:
(nit) a bit hard to read. Think it could have been nicer to do:

```python
if attn_func == 6:
    self.qkv = sefl.decode_qkv
elif attn_func == 7:
    self.qkv = self.prime_qkv
else:
   self.qkv = self.factored_qkv
```

and same for others